### PR TITLE
Change to one endpoint and integrate new frontend options

### PR DESCRIPTION
**Pull request by Lev-Stambler**
- Change backend to use one end point and remove any state, as per issue #50 
- Add options to backend for new frontend options
- Encode options into URL

### CODE REVIEW

File: biogrid/apps/api/src/app/controllers/BiogridController.ts [hunk 1]
```diff
@@ -14,23 +12,9 @@ export class BiogridController extends Controller {
...
-    await runBiogridSimulation();
-  }
-
   @SuccessResponse(200)
-  @Get('/simulation-results')
-  public async GetBiogridSimulationResults(): Promise<
-    BiogridSimulationResults
-  > {
-    return await getSimulationResults();
+  @Post('/')
```
Comment (cwlarson-google) [inline biogrid/apps/api/src/app/controllers/BiogridController.ts]: It seems odd to me to have this "empty" api path.  Even something as simple as "/simulate" would help contextualize

File: biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx [hunk 2]
```diff
@@ -4,15 +4,23 @@ import './input-page.scss';
...
-  const [value, setValue] = useState('');
+function useInput<T>(opts: { default: T }) {
+  const [value, setValue] = useState(opts.default);
+  function valueToType(value: string): T {
+    if (typeof opts.default === 'number') {
+      return (parseInt(value) as unknown) as T;
+    } else if (typeof opts.default === 'boolean') {
+      return (Boolean(value) as unknown) as T;
+    }
+    return (JSON.stringify(value) as unknown) as T;
```
Comment (cwlarson-google) [inline biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx:17]: Can you shorten this to `return JSON.stringify(value) as T;`?

File: biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx [hunk 3]
```diff
@@ -22,92 +30,128 @@ export const InputPage = () => {
   const [startDate, setStartDate] = useState(new Date());
   const [endDate, setEndDate] = useState(new Date());
   const [smallBatteryCells, smallBatteryCellInput] = useInput({
-    type: 'number',
+    default: 0,
```
Comment (cwlarson-google) [inline biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx:33]: If all of these default to 0, can you automatically set that default in the `useInput()` method?

File: biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx [hunk 4]
```diff
@@ -22,92 +30,128 @@ export const InputPage = () => {
...
-      numBuildings,
-      numSolarPanels,
-      townSize,
-
+      smallBatteryCells: smallBatteryCells,
+      largeBatteryCells: largeBatteryCells,
+      numBuildings: numBuildings,
+      numSolarPanels: numSolarPanels,
+      townHeight: townHeight,
+      townWidth: townWidth,
```
Comment (cwlarson-google) [inline biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx]: You shouldn't need to explicitly set the values here since they're the same name

**Review by cwlarson-google — APPROVED** [review]

File: biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx (hunk 2 above)
Comment (Lev-Stambler) [reply biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx:17]: It has to go through unkown or any first before it can be cast as T

File: biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx (hunk 3 above)
Comment (Lev-Stambler) [reply biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx:33]: Oh yeah, earlier there were strings and such, we may add strings in the future though

File: biogrid/apps/api/src/app/services/Biogrid.ts [hunk 5]
```diff
@@ -16,48 +17,52 @@ export interface NewBiogridOpts {
...
-    new Building(10, 4, 3),
-    new Building(10, 1, 2),
-    new Building(10, 3, 1),
-  ];
-  const town = new RuralArea(buildings, 10, 10);
-  biogrid = new Biogrid(town, {
+// TODO, allow users to specify where the buildings are on the grid rather than randomly scatter
+// The issue can be found on https://github.com/googleinterns/step141-2020/issues/59
+export async function simulateNewBiogrid(
+  body: NewBiogridOpts
```
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.ts:41]: nit: rename to "opts"? I think I saw another file in this project that had similar naming convention and want to be consistent

File: biogrid/apps/api/src/app/services/Biogrid.ts [hunk 6]
```diff
@@ -16,48 +17,52 @@ export interface NewBiogridOpts {
...
+): Promise<BiogridSimulationResults> {
+  const buildings = new Array(body.numBuildings).map((v) => {
+    const randomPos = createRandomBuildingPosition(
+      body.townWidth,
+      body.townHeight
+    );
+    return new Building(10, randomPos.x, randomPos.y);
```
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.ts:48]: nit: can you inline comment what the 10 is? 
`return new Building(/* param= */ 10, randomPos.x, randomPox.y);`

File: biogrid/apps/api/src/app/services/Biogrid.ts [hunk 7]
```diff
@@ -16,48 +17,52 @@ export interface NewBiogridOpts {
...
   biogrid.takeAction(action);
-  states.push(biogrid.getJsonGraphDetails());
-  return 'Fake';
-}
-
-export async function getSimulationResults(): Promise<
-  BiogridSimulationResults
-> {
-  // TODO implement
+  statesJson.push(biogrid.getJsonGraphDetails());
```
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.ts:61]: I'm assuming we are pushing this again because you've now taken action and these are different. Can we leave a comment to clarify?

File: biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx (hunk 2 above)
Comment (vitashubin) [reply biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx:17]: but if you're doing parseInt, don't you know that it will be a number?

File: biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx [hunk 8]
```diff
@@ -22,92 +30,128 @@ export const InputPage = () => {
...
-    const body = {
+    function serialize(obj: { [k: string]: unknown }) {
+      const str = [];
+      for (const p in obj)
+        str.push(
+          encodeURIComponent(p) +
+            '=' +
+            encodeURIComponent(
+              (obj[p] as { toString: () => string }).toString()
+            )
```
Comment (vitashubin) [inline biogrid/apps/biogrid-mvp/src/app/pages/input-page/input-page.tsx:62]: can you just use JSON.stringify(obj[p]) here?
